win32/tests/job_object_helper_ut/job_object_helper_ut.c

```diff
@@ -0,0 +1,386 @@
+    REGISTER_GBALLOC_HL_GLOBAL_MOCK_HOOK();
+
+    REGISTER_UMOCK_ALIAS_TYPE(LPMEMORYSTATUSEX, void*);
+    REGISTER_UMOCK_ALIAS_TYPE(JOBOBJECTINFOCLASS, int);
+
+    umock_c_negative_tests_init();
+
+}
+
+TEST_SUITE_CLEANUP(suite_cleanup)
+{
+    umock_c_negative_tests_deinit();
+    umock_c_deinit();
+    real_gballoc_hl_deinit();
+}
+
+TEST_FUNCTION_INITIALIZE(init)
+{
+    umock_c_reset_all_calls();
+}
+
+TEST_FUNCTION_CLEANUP(cleanup)
+{
+}
+
+/*Tests_SRS_JOB_OBJECT_HELPER_18_016: [ job_object_helper_create shall allocate a JOB_OBJECT_HELPER object. ]*/
+/*Tests_SRS_JOB_OBJECT_HELPER_18_023: [ job_object_helper_create shall call GlobalMemoryStatusEx to get the total amount of physical memory in kb. ]*/
+/*Tests_SRS_JOB_OBJECT_HELPER_18_024: [ job_object_helper_create shall call CreateJobObject to create a new job object passing NULL for both lpJobAttributes and lpName. ]*/
+/*Tests_SRS_JOB_OBJECT_HELPER_18_025: [ job_object_helper_create shall call GetCurrentProcess to get the current process handle. ]*/
+/*Tests_SRS_JOB_OBJECT_HELPER_18_026: [ job_object_helper_create shall call AssignProcessToJobObject to assign the current process to the new job object. ]*/
+/*Tests_SRS_JOB_OBJECT_HELPER_18_027: [ job_object_helper_create shall call CloseHandle to close the handle of the current process. ]*/
+/*Tests_SRS_JOB_OBJECT_HELPER_18_019: [ job_object_helper_create shall succeed and return the JOB_OBJECT_HELPER object. ]*/
+TEST_FUNCTION(job_object_helper_create_succeeds)
+{
+    // arrange
+    setup_job_object_helper_create_expectations();
+
+    // act
+    THANDLE(JOB_OBJECT_HELPER) job_object_helper = job_object_helper_create();
+
+    // assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+
+    // cleanup
+    THANDLE_ASSIGN(JOB_OBJECT_HELPER)(&job_object_helper, NULL);
+
+}
+
+/*Tests_SRS_JOB_OBJECT_HELPER_18_018: [ If there are any failures, job_object_helper_create shall fail and return NULL. ]*/
+TEST_FUNCTION(job_object_helper_create_fails)
+{
+    // arrange
+    setup_job_object_helper_create_expectations();
+    umock_c_negative_tests_snapshot();
+
+    for (size_t i = 0; i < umock_c_negative_tests_call_count(); i++)
+    {
+        if (umock_c_negative_tests_can_call_fail(i))
+        {
+            umock_c_negative_tests_reset();
+            umock_c_negative_tests_fail_call(i);
+
+            // act
+            THANDLE(JOB_OBJECT_HELPER) result = job_object_helper_create();
+
+            // assert
+            ASSERT_IS_NULL(result);
+        }
+    }
+
+}
+
+/*Tests_SRS_JOB_OBJECT_HELPER_18_033: [ job_object_helper_dispose shall call CloseHandle to close the handle to the job object. ]*/
+TEST_FUNCTION(job_object_helper_dispose_succeeds)
+{
+    // arrange
+    setup_job_object_helper_create_expectations();
+    THANDLE(JOB_OBJECT_HELPER) job_object_helper = job_object_helper_create();
+    umock_c_reset_all_calls();
+
+    STRICT_EXPECTED_CALL(interlocked_decrement(IGNORED_ARG));
```

Comment:
>interlocked_decrement [](http://example.com/codeflow?start=25&length=21)

where is this coming from?

Reply:
I suppose this is from THANDLE, but we have ways of eliminating that.

Reply:
I was being lazy. Fixed.